lib/srv/db/objects/importer.go

```diff
@@ -0,0 +1,249 @@
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package objects
+
+import (
+	"context"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	dbobjectv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/dbobject/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/utils/retryutils"
+	"github.com/gravitational/teleport/lib/srv/db/common/databaseobjectimportrule"
+	"github.com/gravitational/teleport/lib/srv/db/common/permissions"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/teleport/lib/utils/interval"
+)
+
+type databaseImporter interface {
+	stop()
+}
+
+// singleDatabaseImporter handles importing of objects from a single database.
+type singleDatabaseImporter struct {
+	cfg    Config
+	cancel context.CancelFunc
+
+	database types.Database
+	fetcher  ObjectFetcher
+
+	objects map[string]*objWithExpiry
+}
+
+// objWithExpiry holds an object separate to its expiry time.
+type objWithExpiry struct {
+	obj    *dbobjectv1.DatabaseObject
+	expiry time.Time
+}
+
+type noopImporter struct{}
+
+func (*noopImporter) stop() {}
+
+func startDatabaseImporter(ctx context.Context, cfg Config, database types.Database) (databaseImporter, error) {
+	cfg.Log = cfg.Log.With("database", database.GetName(), "protocol", database.GetProtocol())
+
+	fetcher, err := getObjectFetcher(ctx, database, cfg)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	// nil fetcher indicates unsupported configuration
+	if fetcher == nil {
+		return &noopImporter{}, nil
+	}
+
+	imp := startSingleDatabaseImporter(ctx, cfg, database, fetcher)
+	imp.start(ctx)
+	cfg.Log.InfoContext(ctx, "Successfully started database importer.", "database", database.GetName(), "protocol", database.GetProtocol())
+	return imp, nil
+}
+
+func startSingleDatabaseImporter(ctx context.Context, cfg Config, database types.Database, fetcher ObjectFetcher) *singleDatabaseImporter {
+	cancelCtx, cancel := context.WithCancel(ctx)
+	imp := &singleDatabaseImporter{
+		cfg:      cfg,
+		cancel:   cancel,
+		database: database,
+		fetcher:  fetcher,
+		objects:  make(map[string]*objWithExpiry),
+	}
+	imp.start(cancelCtx)
+	return imp
+}
+
+func (i *singleDatabaseImporter) stop() {
+	i.cancel()
+}
```

Comment:
No, as there may be other importers running, the service may only be down temporarily etc. In general I believe it is better to keep these objects around rather than delete them unnecessarily.

Reply:
sounds good. Let's add a comment on that. i guess that could be part of the future performance improvement.

That said, should `s.cfg.DatabaseObjects.StopImporter` be called in `s.stopProxyingDatabase`? For example, if the `server.Shudown` is called, the importer is not stopped with current impl. IIRC `server.Shudown` can be called during a CA rotation or a systemd reload too.

Reply:
Wouldn't the server shutdown cancel the parent context? I'll check out `stopProxyingDatabase` too.

Reply:
> Wouldn't the server shutdown cancel the parent context? I'll check out `stopProxyingDatabase` too.

good point. as long as it gets stopped during ShutDown, we are good.